### PR TITLE
Support uncertainties as EA Dtype

### PR DESCRIPTION
**Pull request by MichaelTiemannOSC**
In parallel with Pint and Pint-Pandas, changes to support uncertainties as an EA Dtype.

See https://github.com/hgrecco/pint/pull/1615
and https://github.com/hgrecco/pint-pandas/pull/140

- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by jbrockmendel** [comment]
IIUC the problem here is isna not recognizing your custom scalar?  I think the way to address this is directly in #27462/#27825

**Comment by MichaelTiemannOSC** [comment]
Please consider adding the https://github.com/pandas-dev/pandas/milestone/103 label to this pull-request.  There should be no user-visible changes.  The 2.1 milestone contains several enhancements that would further enhance the development versions of Pint and Pint-Pandas which await the changes from this pull request (see https://github.com/hgrecco/pint/pull/1615 and https://github.com/hgrecco/pint-pandas/pull/140).

File: pandas/core/arrays/masked.py [hunk 1]
```diff
@@ -870,7 +870,9 @@ def take(
         # we only fill where the indexer is null
         # not existing missing values
         # TODO(jreback) what if we have a non-na float as a fill value?
-        if allow_fill and notna(fill_value):
+        # NaN with uncertainties is scalar but does not register as `isna`,
+        # so use fact that NaN != NaN
+        if allow_fill and notna(fill_value) and fill_value == fill_value:
```
Comment (jbrockmendel) [inline pandas/core/arrays/masked.py:875]: how would you even get here with uncertainties?  is your EA subclassing BaseMaskedArray?

File: pandas/tests/extension/json/test_json.py [hunk 2]
```diff
@@ -356,9 +356,10 @@ def test_groupby_extension_no_sort(self):
         """
         super().test_groupby_extension_no_sort()
 
-    @pytest.mark.xfail(reason="GH#39098: Converts agg result to object")
-    def test_groupby_agg_extension(self, data_for_grouping):
-        super().test_groupby_agg_extension(data_for_grouping)
+    # GH#39098 is now fixed, so we don't need to XFAIL, nor subclass this test
+    # @pytest.mark.xpass(reason="GH#39098: Converts agg result to object")
+    # def test_groupby_agg_extension(self, data_for_grouping):
+    #     super().test_groupby_agg_extension(data_for_grouping)
```
[the remark this answers is not in the document]
Comment (MichaelTiemannOSC) [reply pandas/tests/extension/json/test_json.py:362]: I think all we need to do is remove the comment and the decorator and let the testsuite do its thing (passing if the test passes, which it should, and failing if the test fails, which it now should not).  I can push that through, but will await other comments first, to save an hour's worth of CI/CD time in case there are other comments.

File: pandas/core/arrays/masked.py (hunk 1 above)
Comment (MichaelTiemannOSC) [reply pandas/core/arrays/masked.py:875]: The EA itself is a very vanilla EA.  The test suite instantiates the EA in myriad ways for test purposes.  In this case the EA gets instantiated as a MaskedArray (which derives from BaseMaskedArray) for testing MaskedArray functionality.  I think the test suite has every right to rigorously instantiate EAs in every way imaginable, which it does.

**Comment by MichaelTiemannOSC** [comment]
My pull request was derailed by a change in pyarrow nightly code caught by CI/CD, likely due to the fact that my branch is now a bit out of date, and not due to the single change I made since passing all tests two days ago.  @mroeschke is active in the pyarrow space and might have suggestions as to how to proceed.  In particular, should I try to sync my fork?  I'm very shy about doing that because in past interactions with git and/or github, sync actions have exploded in my hands in the past, resulting in massively manual merge exercises.  I cannot imagine how much new development pandas has gone through since I forked the repo two weeks ago, so if I need to sync, it would be very helpful to know precisely what buttons to push so that I can make forward progress and not slide massively sideways.

**Comment by MichaelTiemannOSC** [comment]
I also don't understand this complaint from `pylint`:

```
pylint...................................................................Failed
- hook id: pylint
- duration: 793.25s
- exit code: 4

************* Module pandas.tests.extension.json.test_json
pandas/tests/extension/json/test_json.py:[359](https://github.com/pandas-dev/pandas/actions/runs/5553184370/jobs/10141428866?pr=53970#step:12:364):4: W0246: Useless parent or super() delegation in method 'test_groupby_agg_extension' (useless-parent-delegation)
```

The code I un-commented was original code, and it appears to follow precisely the pattern of all other code in the module.  I'm at a loss as to how to proceed because I didn't create the `TestGroupby` class hierarchy and don't know why this particular case is getting singled out.

**Comment by jbrockmendel** [comment]
> The code I un-commented was original code

Again, just remove it and the parent class test will run.

**Comment by MichaelTiemannOSC** [comment]
> > The code I un-commented was original code
> 
> Again, just remove it and the parent class test will run.

Ah, now I understand.  I was afraid to remove a test that should be run, and did not grok that the test would run in the parent class.  I'll re-submit, waiting a few hours in case there's an answer to my pyarrow question.

File: pandas/core/arrays/masked.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/arrays/masked.py:875]: im still confused here.  the EA i see in https://github.com/hgrecco/pint-pandas/pull/140 subclasses ExtensionArray but not BaseMaskedArray, so i don't see how the code change here would affect it.

Comment (MichaelTiemannOSC) [reply pandas/core/arrays/masked.py:875]: OK...I looked more deeply into this when I get back the computer where I did original work and didn't upload some notes to the cloud.  I'm pretty sure it's due to some conditions created when trying to merge two DataFrames that I dummied up in my notes.  I'll create a test case that can go into the Pandas test suite.  It won't depend on uncertainties, but if and when used with uncertainties, it should trigger the path in question.  Should be done by Monday.

Comment (MichaelTiemannOSC) [reply pandas/core/arrays/masked.py:875]: Well, that was a lot more challenging to reconstruct than I expected, but I re-created the condition.  TL;DR: now that I use pd.NA as my NA value for my particular EA, this change that caught your eye is indeed unnecessary and I can revert it.

But, since you asked, here's the path I took initially.  Starting with `PintType` we define this:

```
    @property
    def na_value(self):
        if HAS_UNCERTAINTIES:
            return self.ureg.Quantity(_ufloat_nan, self.units)
        else:
            return self.ureg.Quantity(np.nan, self.units)
```

The idea (before settling on pd.NA, which works without this particular change) was to use a NaN to which we could cast any numerical type (we cannot downcast uncertainties to `float`).

In the test code we set up `data_missing` and `data_missing_for_sorting` to use `ufloat(np.nan, 0)` as the NA value where appropriate.   The test case `test_groupby_agg_extension` uses `data_for_grouping`, which is called without `numeric_dtype` as `None`, and this is where we start going off the rails. 

Here's the `data_for_grouping` code:

```
@pytest.fixture
def data_for_grouping(numeric_dtype, USE_UNCERTAINTIES):
    a = 1.0
    b = 2.0**32 + 1
    c = 2.0**32 + 10
    if USE_UNCERTAINTIES:
        a = a + ufloat(0, 0)
        b = b + ufloat(0, 0)
        c = c + ufloat(0, 0)
        _n = _ufloat_nan
        numeric_dtype = None
    elif numeric_dtype:
        numeric_dtype = dtypemap.get(numeric_dtype, numeric_dtype)
        _n = np.nan
    else:
        _n = pd.NA
    return PintArray.from_1darray_quantity(
        ureg.Quantity(pd.array([b, b, _n, _n, a, a, b, c], dtype=numeric_dtype), ureg.m)
    )
```

The way the test code is run, because `numeric_dtype` is `None` we get:

```
(Pdb) data_for_grouping
<PintArray>
[4294967297.0, 4294967297.0, <NA>, <NA>, 1.0, 1.0, 4294967297.0, 4294967306.0]
Length: 8, dtype: pint[meter]
```

When we execute `expected = df.iloc[[0, 2, 4, 7]]` there comes a point where we come here (`/Users/michael/Documents/GitHub/MichaelTiemannOSC/pandas-dev/pandas/core/internals/managers.py(646)<listcomp>()`):

```
644  	        else:
645  	            new_blocks = [
646  ->	                blk.take_nd(
647  	                    indexer,
648  	                    axis=1,
649  	                    fill_value=(
650  	                        fill_value if fill_value is not None else blk.fill_value
651  	                    ),
```

With `fill_value` as None, we inherit the fill value from the PintArray, which is the uncertain NaN.  Here's the stack trace down from there:

```
  pandas/core/internals/managers.py(645)reindex_indexer()
-> new_blocks = [
  pandas/core/internals/managers.py(646)<listcomp>()
-> blk.take_nd(
  pandas/core/internals/blocks.py(975)take_nd()
-> new_values = algos.take_nd(
  pandas/core/array_algos/take.py(115)take_nd()
-> return arr.take(indexer, fill_value=fill_value, allow_fill=allow_fill)
  pint-pandas/pint_pandas/pint_array.py(523)take()
-> result = take(data, indices, fill_value=fill_value, allow_fill=allow_fill)
  pandas/core/algorithms.py(1307)take()
-> result = take_nd(
  pandas/core/array_algos/take.py(115)take_nd()
-> return arr.take(indexer, fill_value=fill_value, allow_fill=allow_fill)
> pandas/core/arrays/masked.py(879)take()
-> result[fill_mask] = fill_value
```

In this particular case, because none of the indexes are -1, no filling is going to happen.  But should we have things to fill, they'd be filled with the uncertain nan.  And because this code really doesn't want to fill add new NaNs to the array, I put this extra defensive code in.  And that's how we hit this condition.

I'm going to take this extraneous test out (`fill_value == fill_value`).  Does this answer your question?

File: pandas/core/arrays/masked.py [hunk 3]
```diff
@@ -876,6 +876,8 @@ def take(
         # we only fill where the indexer is null
         # not existing missing values
         # TODO(jreback) what if we have a non-na float as a fill value?
+        # NaN with uncertainties is scalar but does not register as `isna`,
+        # so use fact that NaN != NaN
```
Comment (jbrockmendel) [inline pandas/core/arrays/masked.py:880]: is this comment still relevant?

File: pandas/core/groupby/groupby.py [hunk 4]
```diff
@@ -3272,7 +3272,10 @@ def first(x: Series):
                 """Helper function for first item that isn't NA."""
                 arr = x.array[notna(x.array)]
                 if not len(arr):
-                    return np.nan
```
Comment (jbrockmendel) [inline pandas/core/groupby/groupby.py:3275]: could do `x.array._na_value`?  avoids a copy here

File: pandas/core/groupby/ops.py [hunk 5]
```diff
@@ -590,7 +590,7 @@ def nkeys(self) -> int:
 
     def get_iterator(
         self, data: NDFrameT, axis: AxisInt = 0
-    ) -> Iterator[tuple[Hashable, NDFrameT]]:
+    ) -> Iterator[tuple[Hashable, NDFrameT]]:  # Doesn't work with non-hashable EA types
```
Comment (jbrockmendel) [inline pandas/core/groupby/ops.py:593]: pls remove

File: pandas/tests/extension/test_boolean.py [hunk 6]
```diff
@@ -28,7 +28,7 @@
 
 
 def make_data():
-    return [True, False] * 4 + [np.nan] + [True, False] * 44 + [np.nan] + [True, False]
+    return [True, False] * 4 + [pd.NA] + [True, False] * 44 + [pd.NA] + [True, False]
```
Comment (jbrockmendel) [inline pandas/tests/extension/test_boolean.py:31]: why is this necessary?

Comment (MichaelTiemannOSC) [reply pandas/tests/extension/test_boolean.py:31]: I made this change to track the identical changes in test_integer.py and test_floating.py.  The it does not matter whether the NA values in this case are pd.NA, np.nan, or a mixture.  But I did think it better to make the three files consistent.  I'm happy to revert this change if you prefer.

File: pandas/core/arrays/masked.py (hunk 3 above)
Comment (MichaelTiemannOSC) [reply pandas/core/arrays/masked.py:880]: No--I will remove.

File: pandas/tests/extension/test_boolean.py (hunk 6 above)
Comment (mroeschke) [reply pandas/tests/extension/test_boolean.py]: Agreed this should be removed. It's ideal to minimize extraneous changes

File: doc/source/whatsnew/v2.1.0.rst [hunk 7]
```diff
@@ -487,6 +490,7 @@ Conversion
 - Bug in :meth:`DataFrame.info` raising  ``ValueError`` when ``use_numba`` is set (:issue:`51922`)
 - Bug in :meth:`DataFrame.insert` raising ``TypeError`` if ``loc`` is ``np.int64`` (:issue:`53193`)
 - Bug in :meth:`HDFStore.select` loses precision of large int when stored and retrieved (:issue:`54186`)
+-
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.1.0.rst:493]: Could you remove these unrelated changes?

File: doc/source/whatsnew/v2.1.0.rst [hunk 8]
```diff
@@ -640,7 +646,7 @@ Other
 - Bug in :meth:`Series.memory_usage` when ``deep=True`` throw an error with Series of objects and the returned value is incorrect, as it does not take into account GC corrections (:issue:`51858`)
 - Bug in :meth:`period_range` the default behavior when freq was not passed as an argument was incorrect(:issue:`53687`)
 - Fixed incorrect ``__name__`` attribute of ``pandas._libs.json`` (:issue:`52898`)
--
+- Change ``~tests/extension/test_boolean.py`` to use pd.NA instead of np.nan (following similar patterns in ``~tests/extension/test_integer.py`` and ``~tests/extension/test_float.py``), updating :func:`make_data`, :func:`data_missing`, :func:`data_missing_for_sorting`, :func:`data_for_grouping`, :func:`_check_op` (see `PR <https://github.com/pandas-dev/pandas/pull/53970>`_)
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.1.0.rst:649]: Could you remove this? It's not needed

File: doc/source/whatsnew/v2.1.0.rst [hunk 9]
```diff
@@ -160,6 +161,7 @@ Other enhancements
 - :meth:`DataFrame.stack` gained the ``sort`` keyword to dictate whether the resulting :class:`MultiIndex` levels are sorted (:issue:`15105`)
 - :meth:`DataFrame.unstack` gained the ``sort`` keyword to dictate whether the resulting :class:`MultiIndex` levels are sorted (:issue:`15105`)
 - :meth:`DataFrameGroupby.agg` and :meth:`DataFrameGroupby.transform` now support grouping by multiple keys when the index is not a :class:`MultiIndex` for ``engine="numba"`` (:issue:`53486`)
+- :meth:`GroupBy.first` if series contains only NA values (which might be NaN), return the first NA value, else return np.nan (see `PR <https://github.com/pandas-dev/pandas/pull/53970>`_)
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.1.0.rst:164]: Not sure what this is decribing

File: doc/source/whatsnew/v2.1.0.rst [hunk 10]
```diff
@@ -141,6 +141,7 @@ Other enhancements
 - :meth:`MultiIndex.sortlevel` and :meth:`Index.sortlevel` gained a new keyword ``na_position`` (:issue:`51612`)
 - :meth:`arrays.DatetimeArray.map`, :meth:`arrays.TimedeltaArray.map` and :meth:`arrays.PeriodArray.map` can now take a ``na_action`` argument (:issue:`51644`)
 - :meth:`arrays.SparseArray.map` now supports ``na_action`` (:issue:`52096`).
+- :meth:`arrays.BaseMaskedArray.take` handle non-na float as fill value (triggered by ``ufloat`` NaN from ``uncertainties`` package) (see `PR <https://github.com/pandas-dev/pandas/pull/53970>`_)
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.1.0.rst:144]: Could you describe this in terms of a public API? (BaseMaskedArray is not public)

Also ```(:issue:`53970`)``` is the appropriate citation

File: doc/source/whatsnew/v2.1.0.rst (hunk 9 above)
Comment (MichaelTiemannOSC) [reply doc/source/whatsnew/v2.1.0.rst:164]: This is describing a change of behavior of the ``GroupBy.first`` helper function to not assume that np.nan is a valid NA value for the EA in question.  The idea of the change (which, if valid, should also be applied to ``GroupBy.last``) is that if we cannot find any non-NA values in the array, but if there are NA values, the return value should be the first (or last) NA value in the array.  I did this because I thought that my EA could only deal with NaN values that were UFloats (having a nominal value and a std_dev value).  But as I traced through the code and observed things more closely, I see that the EA does have tolerance for np.nan as an element among UFloats, giving np.nan as the nominal value and zero as the std_dev.  So I will remove this change (by being less strict in some of my own assumptions in checking EA consistency).

File: pandas/core/groupby/groupby.py (hunk 4 above)
Comment (MichaelTiemannOSC) [reply pandas/core/groupby/groupby.py:3275]: One better...I can revert to original Pandas code.

File: doc/source/whatsnew/v2.1.0.rst (hunk 10 above)
Comment (MichaelTiemannOSC) [reply doc/source/whatsnew/v2.1.0.rst:144]: I removed the code to which this documentation was referring: https://github.com/pandas-dev/pandas/pull/53970/commits/b8083a9a2559d22e41991a1d4c41fa266b64dfd3#diff-dddb661834ccfb6fde724041184559470d50f52e56719bdfa9c0d5e7922f0a51

I will revert this patch.

**Comment by MichaelTiemannOSC** [comment]
This PR is obsolete now.
